src/types/type_param.rs

```diff
@@ -84,35 +103,103 @@ pub enum TypeArg {
         arg: CustomTypeArg,
     },
     /// Instance of [TypeParam::List] or [TypeParam::Tuple], defined by a
-    /// sequence of arguments.
+    /// sequence of elements.
     Sequence {
         #[allow(missing_docs)]
-        args: Vec<TypeArg>,
+        elems: Vec<TypeArg>,
     },
     /// Instance of [TypeParam::Extensions], providing the extension ids.
     Extensions {
         #[allow(missing_docs)]
         es: ExtensionSet,
     },
+    /// Variable (used in type schemes only) - see [TypeArg::new_var_use]
+    Variable {
+        #[allow(missing_docs)]
+        v: TypeArgVariable,
```

Comment:
Is there a reason why you declare the `TypeArgVariable` struct instead of just having `idx` and `chached_decl` here?

Reply:
One of Rust's annoyances is that to match on the enum you have to make it `pub`, which makes the whole structure (all variants and their fields) public - I mean, it has to, really - and then anyone can create instances like `TypeArg::Variable(2, TypeParam::Type(TypeBound::Eq))`. This is a malformed instance, as it needs to be represented as `TypeArg::Type(Type::new_variable(2, TypeBound::Eq))`.

I think you'd figured that out already. Of course we could detect malformed instances only in `validate` and that would still be OK - this just prevents them from ever being constructed i.e. detects them much eariler.